pkg/infrastructure/gcp/clusterapi/iam.go

```diff
@@ -102,48 +106,56 @@ func AddServiceAccountRoles(ctx context.Context, projectID, serviceAccountID str
 		return fmt.Errorf("failed to create resourcemanager service: %w", err)
 	}
 
-	policy, err := getPolicy(ctx, service, projectID)
-	if err != nil {
-		return err
+	backoff := wait.Backoff{
+		Duration: 2 * time.Second,
+		Jitter:   1.0,
+		Steps:    5,
 	}
-
-	member := fmt.Sprintf("serviceAccount:%s", serviceAccountID)
-	for _, role := range roles {
-		err = addMemberToRole(policy, role, member)
-		if err != nil {
-			return fmt.Errorf("failed to add role %s to %s: %w", role, member, err)
+	// Get and set the policy in a backoff loop and return errors if any.
+	return wait.ExponentialBackoffWithContext(ctx, backoff, func(ctx context.Context) (bool, error) {
+		policy, err := getPolicy(ctx, service, projectID)
+		if isQuotaExceededError(err) {
+			logrus.Debugf("Failed to get IAM policy, retrying after backoff")
+			return false, nil
+		} else if err != nil {
+			return false, fmt.Errorf("failed to get IAM policy, unexpected error: %w", err)
 		}
-	}
 
-	err = setPolicy(ctx, service, projectID, policy)
-	if err != nil {
-		return err
-	}
+		member := fmt.Sprintf("serviceAccount:%s", serviceAccountID)
+		for _, role := range roles {
+			err = addMemberToRole(policy, role, member)
+			if err != nil {
+				return false, fmt.Errorf("failed to add role %s to %s: %w", role, member, err)
+			}
+		}
```

Comment:
From what I can see, `addMemberToRole` never fails and it doesn't do any API calls. So once you have the `policy` from the `getPolicy` call, you can run this only once and it doesn't need to be part of the retry loop, right?

Reply:
Reconfirmed that doing the Get is required if the Set fails, otherwise the wait fails with
`failed to set IAM policy, unexpected error: googleapi: Error 409: There were concurrent policy changes. Please retry the whole read-modify-write with exponential backoff. The request's ETag '\\007\\006\\036\\022Y\\211\\3532' did not match the current policy's ETag '\\007\\006\\036\\022\\334\\353\\212\\350'., aborted"`

Will add a comment to that affect.